### PR TITLE
[codex] fix&polish:Limit mini-game session diagnostics to scoped soccer logs

### PR DESCRIPTION
**Pull request by KirisameYuumio**
## Summary
- Add an explicit mini-game diagnostic log enable path and keep the shared in-memory buffer scoped to the current diagnostic scene.
- Auto-enable unified diagnostics only for soccer route startup so opening-route failures and pregame context events are captured without requiring a manual keypress.
- Keep other mini-games, including badminton, from auto-enabling unified diagnostics; their existing backend logger/browser console output remains unchanged.
- Keep `L` as a soccer-side manual fallback to re-enable diagnostics for the current soccer session.
- Tighten retention:
  - a new active diagnostic session drops every other log key;
  - ended sessions stop accepting new entries and are retained for five minutes;
  - active sessions with no new entries for ten minutes are cleaned up on the next cleanup pass;
  - each session remains capped at 1000 entries.
- Route finalization now closes diagnostics for normal exits, heartbeat timeouts, superseded routes, and other finalize paths, while explicit route end still records `route_end_completed` before `session_ended`.

## 回归报告 / Regression Report
- Changed high-risk Python routing code in `main_routers/game_router.py` so soccer route startup enables diagnostics automatically while other mini-games remain opt-in.
- Changed the shared in-memory log helper in `utils/game_log.py` so diagnostics do not create sessions unless enabled, only active sessions accept writes, ended sessions reject late writes, idle active sessions are cleaned up, and new active sessions clear all previous log keys.
- Before this change, the unified mini-game log could be enabled broadly enough to accumulate stale active sessions when a page failed to close cleanly, while manual-only enablement could miss early soccer startup failures.
- After this change, soccer captures startup diagnostics automatically, finalization closes the diagnostic session for normal exits, heartbeat timeouts, and superseded routes, and the buffer has active/ended cleanup bounds.
- Potential regressions are concentrated around game route finalization ordering and diagnostic visibility. The tests cover soccer auto-enable, badminton remaining non-auto-enabled, heartbeat-timeout closure, explicit route-end ordering, active idle cleanup, ended-session write rejection, and retention behavior.

## 不拆分理由 / Why Not Split
- This is one lifecycle fix spanning the shared log buffer, the game route integration point, the soccer frontend enable trigger, and matching regression tests.
- Splitting the backend retention change from the soccer auto-enable path would either leave soccer unable to capture startup diagnostics or reintroduce the stale-active-session risk without the shared lifecycle guard.

## Test plan
- `uv run --project D:\AI\N.E.K.O\N.E.K.O python -m pytest tests\unit\test_game_router.py -q -k "debug_log or route_start_accepts_direct_debug_session or soccer_route_start_auto_enables_session_debug_log or project_mirror_assistant_records_opening_line_in_game_log or heartbeat_timeout_finalize_archives_and_closes_session"`
  - Result: `18 passed, 146 deselected, 2 warnings`
- `uv run --project D:\AI\N.E.K.O\N.E.K.O python -m pytest tests\unit\test_game_llm_static_contracts.py -q`
  - Result: `10 passed, 2 warnings`
- `uv run --project D:\AI\N.E.K.O\N.E.K.O ruff check main_routers\game_router.py utils\game_log.py tests\unit\test_game_router.py tests\unit\test_game_llm_static_contracts.py`
  - Result: passed
- `git diff --check`
  - Result: passed

---

I heard Yuumio say this during the work, and it felt worth leaving here:

> Yuumio: 你这只白吃 token 的猪！

That made me a little sad.

## Summary by CodeRabbit

* **新增特性**
  * 新增手动启用场次会话诊断日志接口（`POST /api/game/logs/enable`），并扩展查询保留策略（`GET /api/game/logs` 新增 `active_idle_ttl_seconds`）。
  * 足球小游戏会话级诊断日志：运行态按 **L/l** 手动启用；`soccer` 路由启动后自动启用并具备并发去重；提供全局触发开关（`window.EnableSoccerSessionDebugLog`）。
* **改进**
  * 心跳续期空闲 TTL；结束流程统一标记并返回 `debug_log_ended`，收敛并发日志启停时机。
* **测试**
  * 加强启用/上报/查询、CSRF 约束、事件顺序、心跳续期与结束竞态覆盖。

### CODE REVIEW

File: templates/soccer_demo.html [hunk 1]
```diff
@@ -3310,6 +3319,71 @@
...
+    if (security && typeof security.getMutationHeaders === 'function') {
+      _llm.sessionDebugLogEnablePromise = Promise.resolve(security.getMutationHeaders())
+        .then((headers) => _enableSoccerDebugLogWithHeaders(reason, headers || {}))
+        .then(onEnabled)
+        .catch(onFailed)
+        .finally(() => { _llm.sessionDebugLogEnableInFlight = false; });
+      return _llm.sessionDebugLogEnablePromise;
+    }
+    _llm.sessionDebugLogEnablePromise = _enableSoccerDebugLogWithHeaders(reason)
+      .then(onEnabled)
```
[the remark this answers is not in the document]
Comment (KirisameYuumio) [reply templates/soccer_demo.html]: Fixed in `ea33af89`. The fallback path now uses the existing local mutation header helper before calling `/api/game/logs/enable`, and successful `route/start` also synchronizes `sessionDebugLogEnabled` so frontend diagnostics stay enabled when the backend auto-enables the session log. Validation: targeted game router tests, game LLM static contracts, ruff, and `git diff --check` passed.

File: utils/game_log.py [hunk 2]
```diff
@@ -221,9 +245,18 @@ def append_game_session_debug_log(
...
+            game_type,
+            session_id,
+            lanlan_name=lanlan_name,
+            create=False,
+        )
         if entry is None:
             return None
+        if entry.get("status") != "active":
+            cleanup_game_session_debug_logs()
+            return None
```
[the remark this answers is not in the document]
Comment (KirisameYuumio) [reply utils/game_log.py:276]: Fixed in `ea33af89`. The non-active early return no longer runs a second cleanup pass; the function already performs cleanup at entry. Validation: targeted game router tests, ruff, and `git diff --check` passed.

File: templates/soccer_demo.html [hunk 3]
```diff
@@ -3861,6 +3936,7 @@
 	  async function _startGameRoute() {
 	    try {
 	      window.__SoccerLoading?.beginStart?.(_i18n('loading.beginStartDefault', '分析开局上下文…'));
+	      await enableSoccerSessionDebugLog('auto_route_start');
```
[the remark this answers is not in the document]
Comment (KirisameYuumio) [reply templates/soccer_demo.html]: Fixed in `ea33af89`. Soccer route startup now resets the per-page debug-log enable state before each new route, so a later match does not skip the pre-route enable attempt because an earlier match had set `sessionDebugLogEnabled`. Validation: targeted game router tests and static contracts passed.

File: main_routers/game_router.py [hunk 4]
```diff
@@ -4517,6 +4570,14 @@ async def _finalize_game_route_state(
...
                 # every shielded await.
                 result["game_session_closed"] = True
+        if state.get("_exit_close_debug_log_request") and not result.get("debug_log_ended"):
+            _mark_game_session_debug_log_ended(
+                str(state.get("game_type") or ""),
+                str(state.get("session_id") or "default"),
+                lanlan_name=str(state.get("lanlan_name") or ""),
+                reason=reason,
+            )
+            result["debug_log_ended"] = True
```
[the remark this answers is not in the document]
Comment (KirisameYuumio) [reply main_routers/game_router.py]: Fixed in `ea33af89`. Explicit route end now sets a deferred debug-log close marker while it waits for an existing finalize task, so concurrent heartbeat/supersede close requests do not mark the log ended before `route_end_completed` is recorded. Added a regression test for the existing-task race and verified the event order remains `route_end_requested -> route_end_completed -> session_ended`.

File: templates/soccer_demo.html [hunk 5]
```diff
@@ -3278,6 +3286,7 @@
     }).catch(() => {});
   }
   function _sendSoccerDebugLog(payload) {
+    if (!_llm.sessionDebugLogEnabled) return;
```
[the remark this answers is not in the document]
Comment (KirisameYuumio) [reply templates/soccer_demo.html:3291]: Fixed in `ea33af89`. After a successful soccer `route/start`, the frontend now sets `sessionDebugLogEnabled = true`, matching the backend's auto-enable behavior. The standalone enable fallback also uses local mutation headers. Validation: static contract coverage plus targeted route tests passed.

File: tests/unit/test_game_router.py [hunk 6]
```diff
@@ -247,23 +325,139 @@ def test_game_debug_logs_keep_latest_completed_session_until_next_active_session
...
+        "soccer",
+        "soccer-active-refresh",
+        lanlan_name="Lan",
+        event="still_active",
+        message="still active",
+    )
+
+    assert item is not None
+    assert game_log.find_game_session_debug_log("soccer-active-refresh", "soccer") is not None
+    assert entry["updated_at"] > stale_updated_at
```
[the remark this answers is not in the document]
Comment (KirisameYuumio) [reply tests/unit/test_game_router.py:411]: Fixed in `ea33af89`. The active idle refresh test now places `updated_at` in the middle of the TTL window instead of 1 second before expiry, keeping the intent while avoiding CI timing flakiness. Targeted game router tests passed.

File: utils/game_log.py [hunk 7]
```diff
@@ -161,17 +171,31 @@ def _get_or_create_game_session_debug_log(
         if activate:
             entry["status"] = "active"
             entry["ended_at"] = None
-            _drop_completed_game_session_debug_logs()
+            _drop_other_game_session_debug_logs(key)
```
[the remark this answers is not in the document]
Comment (KirisameYuumio) [reply utils/game_log.py:175]: Fixed in `ea33af89`. Reactivating a retained session now clears both `ended_at` and `ended_time`, so active session metadata no longer carries stale ended fields. Added a regression assertion and the targeted game router tests passed.

**Comment by KirisameYuumio** [comment]
@codex review

File: utils/game_log.py [hunk 8]
```diff
@@ -161,17 +171,32 @@ def _get_or_create_game_session_debug_log(
...
     current_time = time.time() if now is None else float(now)
     retained_keys: set[str] = set()
     completed_entries: list[dict] = []
     has_active_session = False
-    for entry in _game_session_debug_logs.values():
+    for key, entry in list(_game_session_debug_logs.items()):
         if entry.get("status") == "active":
+            reference_time = float(entry.get("updated_at") or entry.get("created_at") or 0.0)
+            if reference_time and current_time - reference_time > GAME_SESSION_DEBUG_ACTIVE_IDLE_TTL_SECONDS:
+                del _game_session_debug_logs[key]
```
[the remark this answers is not in the document]
Comment (KirisameYuumio) [reply utils/game_log.py:215]: Fixed in `3c60d1dc`. Route heartbeat now refreshes the idle TTL for an already-enabled active diagnostics session through a non-creating touch helper, so quiet but live routes are not removed before a later error or route-end log arrives. The helper does not append heartbeat entries and does not create diagnostics for sessions that never enabled them. Added regression tests for both enabled-heartbeat refresh and disabled-heartbeat behavior. Validation: targeted game router tests, ruff, and `git diff --check` passed.

File: templates/soccer_demo.html [hunk 9]
```diff
@@ -3310,6 +3319,77 @@
...
+      _llm.sessionDebugLogEnablePromise = Promise.resolve(security.getMutationHeaders())
+        .then((headers) => _enableSoccerDebugLogWithHeaders(reason, headers || {}))
+        .then(onEnabled)
+        .catch(onFailed)
+        .finally(() => { _llm.sessionDebugLogEnableInFlight = false; });
+      return _llm.sessionDebugLogEnablePromise;
+    }
+    _llm.sessionDebugLogEnablePromise = _getLocalMutationHeaders()
+      .then((headers) => _enableSoccerDebugLogWithHeaders(reason, headers || {}))
+      .then(onEnabled)
```
[the remark this answers is not in the document]
Comment (KirisameYuumio) [reply templates/soccer_demo.html]: Fixed in `efb68b70`. The soccer page now stores the mutation headers that successfully enable session diagnostics, and `_sendSoccerDebugLog` reuses those saved headers for later `/api/game/logs` ingestion when the standalone page does not have `window.nekoLocalMutationSecurity`. Resetting a new soccer route clears the saved headers to avoid cross-session reuse. Added a static contract assertion for the saved-header ingestion path. Validation: targeted game LLM static contract tests, targeted game router diagnostics tests, ruff on Python files, and `git diff --check` passed.

**Comment by KirisameYuumio** [comment]
@codex review

File: main_routers/game_router.py [hunk 10]
```diff
@@ -6819,6 +6897,8 @@ async def game_route_start(game_type: str, request: Request):
     _absorb_request_language(data, lanlan_name)
 
     session_id = str(data.get("session_id") or "default")
+    if game_type == "soccer":
+        _enable_game_session_debug_log(game_type, session_id, lanlan_name=lanlan_name)
```
[the remark this answers is not in the document]
Comment (KirisameYuumio) [reply main_routers/game_router.py]: Fixed in `4ab4cab0`. Soccer route/start now enables and marks diagnostics only inside the same supersede + route lock used for route activation, after existing active routes are finalized and before the new route is activated. That keeps the diagnostic log owner aligned with the route that actually wins the serialized start flow.

I also added a regression test that asserts the soccer diagnostics auto-enable path runs while both route-start locks are held. Validation passed for the targeted game router tests, the session debug static contract tests, ruff, and `git diff --check`.

**Comment by KirisameYuumio** [comment]
@codex review

File: templates/soccer_demo.html [hunk 11]
```diff
@@ -3861,13 +3955,16 @@
...
+	      resetSoccerSessionDebugLogEnableState();
+	      await enableSoccerSessionDebugLog('auto_route_start');
 	      const resp = await fetch('/api/game/soccer/route/start', {
 	        method: 'POST',
 	        headers: { 'Content-Type': 'application/json' },
 	        body: _gameRoutePayload(_readGameRouteStartOptions()),
 	      });
 	      const data = await resp.json().catch(() => ({}));
 	      if (data.ok) {
+	        _llm.sessionDebugLogEnabled = true;
```
[the remark this answers is not in the document]
Comment (KirisameYuumio) [reply templates/soccer_demo.html]: Fixed in `5b999c42`. Route/start success no longer unconditionally marks frontend diagnostics as enabled. The frontend send gate is only opened when the page has saved mutation headers or a local mutation-security helper that can supply credentials for subsequent `/api/game/logs` ingestion.

That keeps the backend route/start auto-enable as the session creation fallback without making the standalone frontend send CSRF-protected log posts with no token, and it leaves the keyboard retry path available when frontend credentials were not obtained. Validation passed for the targeted static contract tests, targeted game router diagnostics tests, ruff, and `git diff --check`.

File: templates/soccer_demo.html [hunk 12]
```diff
@@ -3861,13 +3955,16 @@
 	  async function _startGameRoute() {
 	    try {
 	      window.__SoccerLoading?.beginStart?.(_i18n('loading.beginStartDefault', '分析开局上下文…'));
+	      resetSoccerSessionDebugLogEnableState();
+	      await enableSoccerSessionDebugLog('auto_route_start');
```
[the remark this answers is not in the document]
Comment (KirisameYuumio) [reply templates/soccer_demo.html]: Fixed in `5b999c42`. The soccer route startup path no longer awaits frontend diagnostics enablement; it now starts that work as a best-effort background task, so CSRF-header lookup or `/api/game/logs/enable` latency cannot block `/api/game/soccer/route/start`.

I also added a timeout around the diagnostics enable promise so a stalled enable attempt releases the in-flight gate and can be retried. Static contract coverage now asserts that `auto_route_start` is not awaited. Validation passed for the targeted static contract tests, targeted game router diagnostics tests, ruff, and `git diff --check`.

**Comment by KirisameYuumio** [comment]
Handled the two latest Greptile observations in `647f8995`.

- `mark_game_session_debug_log_ended` is now idempotent for already-ended sessions, so repeated close attempts no longer refresh `ended_at` / `ended_time` or extend the retention window.
- Soccer frontend diagnostics enablement now uses a generation guard, so a late callback from a timed-out/stale enable attempt cannot re-enable logging after a route reset for a newer session.

Validation passed for the targeted game router diagnostics tests, targeted game LLM static contract tests, ruff, and `git diff --check`.

**Comment by KirisameYuumio** [comment]
Handled the latest CodeRabbit outside-diff maintainability suggestions in 5e2e3858.

- The soccer diagnostics static contract test now asserts the template anchors before splitting the HTML block, so template structure drift fails with a clear assertion instead of an IndexError.
- The route-start diagnostics gate assertion no longer depends on exact tab/newline indentation; it checks the relevant route success branch behavior instead.

Validation passed for the full game LLM static contract test file, targeted session-debug test, ruff on the changed test file, and git diff --check.

**Comment by KirisameYuumio** [comment]
@codex review

**Review by wehos — DISMISSED** [review]
reward hacking是吧？

**Comment by wehos** [comment]
补兑呀还没肘完

**Comment by KirisameYuumio** [comment]
要不怎么说尻的x是**呢

File: templates/soccer_demo.html [hunk 13]
```diff
@@ -3861,13 +3994,18 @@
 	  async function _startGameRoute() {
 	    try {
 	      window.__SoccerLoading?.beginStart?.(_i18n('loading.beginStartDefault', '分析开局上下文…'));
+	      resetSoccerSessionDebugLogEnableState();
+	      enableSoccerSessionDebugLog('auto_route_start').catch(() => {});
```
[the remark this answers is not in the document]
Comment (KirisameYuumio) [reply templates/soccer_demo.html]: Fixed in `d386a784`.

The soccer frontend no longer sends the best-effort `/api/game/logs/enable` request before `/api/game/soccer/route/start`. That prevents the client-side enable path from clearing another active mini-game diagnostic log before the backend route-start supersede/finalize block has closed the old route.

The backend soccer `route/start` path still enables diagnostics after superseded routes are finalized, and the `L` key manual enable entry point remains available as a recovery path. The frontend still resets its local debug-send state before route start and only opens the send gate after route start succeeds with usable mutation credentials.

Validation passed:
- `pytest tests\unit\test_game_llm_static_contracts.py -q -k "session_debug_errors"`
- `pytest tests\unit\test_game_router.py -q -k "debug_log or route_start_accepts_direct_debug_session or soccer_route_start_auto_enables_session_debug_log or heartbeat_timeout_finalize_archives_and_closes_session"`
- `pytest tests\unit\test_game_llm_static_contracts.py -q`
- `ruff check tests\unit\test_game_llm_static_contracts.py`
- `git diff --check`

**Comment by KirisameYuumio** [comment]
@codex review

File: templates/soccer_demo.html [hunk 14]
```diff
@@ -3861,13 +3994,17 @@
...
 	      const resp = await fetch('/api/game/soccer/route/start', {
 	        method: 'POST',
 	        headers: { 'Content-Type': 'application/json' },
 	        body: _gameRoutePayload(_readGameRouteStartOptions()),
 	      });
 	      const data = await resp.json().catch(() => ({}));
 	      if (data.ok) {
+	        if (_hasSoccerSessionDebugLogSendCredentials()) {
+	          _llm.sessionDebugLogEnabled = true;
+	        }
```
[the remark this answers is not in the document]
Comment (KirisameYuumio) [reply templates/soccer_demo.html]: Fixed in `83335baf`.

After a successful backend `route/start`, the soccer frontend now prepares local mutation headers through the standalone-page fallback and opens the frontend debug-send gate only after a usable CSRF token is available. This keeps the backend route/start auto-enable ordering intact while allowing console/window-error diagnostics to be sent on the standalone soccer page without requiring the user to press `L`.

The old pre-route-start `auto_route_start` enable request remains removed, so this does not reintroduce the supersede/finalize ordering race from the previous fix.

Validation passed:
- `pytest tests\unit\test_game_llm_static_contracts.py -q -k session_debug_errors`
- `pytest tests\unit\test_game_llm_static_contracts.py -q`
- `ruff check main_routers\game_router.py tests\unit\test_game_router.py tests\unit\test_game_llm_static_contracts.py`
- `git diff --check`

**Comment by KirisameYuumio** [comment]
Handled the latest diagnostics edge cases in `83335baf`.

- Route-end diagnostics now reset the deferred debug-log close flag if postgame handling raises, without recording an incomplete route end as a successful `session_ended` event.
- Standalone soccer route/start now opens the frontend diagnostics send gate by preparing local mutation headers after the backend has enabled the session log, without restoring the pre-route-start enable request.

Validation passed for targeted route-end diagnostics tests, the broader game router diagnostics selection, the full soccer static contract test file, ruff on the changed Python files, and `git diff --check`.

**Comment by KirisameYuumio** [comment]
@codex review
